src/test/resources/com/jpinpoint/perf/lang/kotlin/ruleset/common/xml/AvoidImplicitlyRecompilingRegex.xml

```diff
@@ -0,0 +1,361 @@
+        val e = context1.matches(arg.toRegex()) // good, dynamic
+        val f = context1.matches("\\w".toRegex()) // bad, no matter it is short
+        val local = arg
+        val g = context1.matches(method(arg).toRegex()) // good, dynamic
+        val h = context1.matches(method2().toRegex()) // good, dynamic
+        val i = context1.matches(local.toRegex()) // good, dynamic
+    }
+
+    fun method(arg: String): String {
+        return arg
+    }
+
+    fun method2(): String {
+        return arg
+    }
+}
+     ]]></code>
+    </test-code>
+    <test-code>
+        <description>violation: Avoid implicit recompiling of regular expressions - non toRegex</description>
+        <expected-problems>8</expected-problems>
+        <expected-linenumbers>16,24,25,26,32,33,34,34</expected-linenumbers>
+        <!-- TODO: missing is ,37 -->
+        <!--
+           (: also accept var based on FileSystems.getDefault(), now a match on 'simple' identifier (e.g. var) is accepted :)
+               or (ancestor::PostfixUnaryExpression/PrimaryExpression/SimpleIdentifier/T-Identifier[
+               @Text = ancestor::FunctionBody//SimpleIdentifier/T-Identifier[@Text and ../../../../PropertyDeclaration//Expression//SimpleIdentifier/T-Identifier[@Text='FileSystems']]]
+           )
+        -->
+        <code><![CDATA[
+import java.io.File
+import java.nio.file.FileSystems
+import java.nio.file.Path
+import java.util.regex.Pattern
+
+class Foo {
+
+    companion object {
+        private const val PAT_STRING = "(?=\\p{Lu})"
+        private const val PAT_STRING_SHORT = "/$1"
+        private val path: Path = File("fileName").toPath()
+        private const val FREE_TEXT_PATTERN = "[A-Z\\s]+"
+
+        // line 14
+        fun isValid_violate1x(text: String): Boolean {
+            return text.length > 0 && !Pattern.matches(FREE_TEXT_PATTERN, text.trim { it <= ' ' }) // bad line 109, use Kotlin String matches
+        }
+    }
+
+    // Pattern.matches
+    // line 21
+    fun violatingPatternMatches3x(action: String?) {
+        val b =
+            Pattern.matches("(?=\\p{Lu})", action) // bad: use Kotlin String matches with precompiled Regex
+        val c = Pattern.matches(PAT_STRING, action) // bad: use Kotlin String matches with precompiled Regex
+        val d = Pattern.matches(context1, action) // bad: use Kotlin String matches with toRegex()
+    }
+
+    // line 29
+    fun violatingPathMatcher4x(action: String?) {
+        var p = FileSystems.getDefault()
+            .getPathMatcher("(?=\\p{Lu})") // bad: violation, should not be inside method, make static final
+        FileSystems.getDefault().getPathMatcher(PAT_STRING) // bad: violation
+        FileSystems.getDefault().getPathMatcher(dynPatString).matches(path) // bad (2x):  field should be PathMatcher
+
+        val fs = FileSystems.getDefault()
+        p = fs.getPathMatcher(PAT_STRING) // bad: violation
```

Comment:
To match this line, the last part of xpath expression can be:

```
(: check if java.nio.file.FileSystems or java.nio.file.Path imports exists as filter for same named classes in other packages :)
(: TODO is this a bug: having to use T-FILE for 'file' :)
//ImportHeader[
    (.//T-Identifier[@Text='java']
        and .//T-Identifier[@Text='nio'] and .//T-FILE[@Text='file'] and .//T-Identifier[@Text='FileSystems' or @Text='Path'] or .//T-MULT)
  ][1]/ancestor::KotlinFile
//FunctionDeclaration//PostfixUnarySuffix//T-Identifier[(@Text='getPathMatcher' or @Text='matches')
    and (ancestor::PostfixUnaryExpression/PrimaryExpression/SimpleIdentifier/T-Identifier[@Text='FileSystems'])
    (: not if first param is function parameter :)
    and not(ancestor::PrefixUnaryExpression//CallSuffix/ValueArguments/ValueArgument//SimpleIdentifier/T-Identifier[
       @Text = ancestor::FunctionDeclaration//Parameter/SimpleIdentifier/T-Identifier/@Text])
    (: not param in String template in function param :)
    and not(ancestor::PrefixUnaryExpression//CallSuffix/ValueArguments/ValueArgument//LineStringContent/T-LineStrRef[
       @Text = ancestor::FunctionDeclaration//Parameter/SimpleIdentifier/T-Identifier/concat('$', @Text)])
    or (ancestor::PostfixUnaryExpression/PrimaryExpression//T-Identifier[@Text=
        ancestor::Statement/preceding-sibling::Statement//VariableDeclaration[../Expression//T-Identifier[@Text='FileSystems']]//T-Identifier/@Text])
]
```

The last 3 full lines are new.
Only thing is that it matches not just on 'getPathMatcher' but also on 'PAT_STRING' which I don't understand.